File: examples/attitude/rotation_matrix_operations.py
# /// script
# dependencies = ["brahe"]
# ///

"""
Demonstrates rotation matrix operations including matrix multiplication
and vector transformations.
"""

import brahe as bh
import numpy as np

# Create two rotation matrices
# 90° rotation about X-axis
rm_x = bh.RotationMatrix(1.0, 0.0, 0.0, 0.0, 0.0, -1.0, 0.0, 1.0, 0.0)

# 90° rotation about Z-axis
rm_z = bh.RotationMatrix(0.0, -1.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 1.0)

print("Rotation matrix X (90° about X):")
print(f"  [{rm_x.r11:.3f}, {rm_x.r12:.3f}, {rm_x.r13:.3f}]")
print(f"  [{rm_x.r21:.3f}, {rm_x.r22:.3f}, {rm_x.r23:.3f}]")
print(f"  [{rm_x.r31:.3f}, {rm_x.r32:.3f}, {rm_x.r33:.3f}]")

print("\nRotation matrix Z (90° about Z):")
print(f"  [{rm_z.r11:.3f}, {rm_z.r12:.3f}, {rm_z.r13:.3f}]")
print(f"  [{rm_z.r21:.3f}, {rm_z.r22:.3f}, {rm_z.r23:.3f}]")
print(f"  [{rm_z.r31:.3f}, {rm_z.r32:.3f}, {rm_z.r33:.3f}]")

# Matrix multiplication (compose rotations)
# Apply rm_x first, then rm_z
rm_composed = rm_z * rm_x
print("\nComposed rotation (X then Z):")
print(f"  [{rm_composed.r11:.3f}, {rm_composed.r12:.3f}, {rm_composed.r13:.3f}]")
print(f"  [{rm_composed.r21:.3f}, {rm_composed.r22:.3f}, {rm_composed.r23:.3f}]")
print(f"  [{rm_composed.r31:.3f}, {rm_composed.r32:.3f}, {rm_composed.r33:.3f}]")

# Transform a vector using rotation matrix
# Rotate vector [1, 0, 0] by 90° about Z-axis using matrix multiplication
R_z = rm_z.to_matrix()  # Get 3x3 numpy array
vector = np.array([1.0, 0.0, 0.0])
rotated = R_z @ vector  # Matrix-vector multiplication
print("\nVector transformation:")
print(f"  Original: [{vector[0]:.3f}, {vector[1]:.3f}, {vector[2]:.3f}]")
print(f"  Rotated:  [{rotated[0]:.3f}, {rotated[1]:.3f}, {rotated[2]:.3f}]")

# Transform another vector
vector2 = np.array([0.0, 1.0, 0.0])
rotated2 = R_z @ vector2
print(f"\n  Original: [{vector2[0]:.3f}, {vector2[1]:.3f}, {vector2[2]:.3f}]")
print(f"  Rotated:  [{rotated2[0]:.3f}, {rotated2[1]:.3f}, {rotated2[2]:.3f}]")

# Expected output:
# Rotation matrix X (90° about X):
#   [1.000, 0.000, 0.000]
#   [0.000, 0.000, -1.000]
#   [0.000, 1.000, 0.000]
#
# Rotation matrix Z (90° about Z):
#   [0.000, -1.000, 0.000]
#   [1.000, 0.000, 0.000]
#   [0.000, 0.000, 1.000]
#
# Composed rotation (X then Z):
#   [0.000, 0.000, 1.000]
#   [1.000, 0.000, 0.000]
#   [0.000, 1.000, 0.000]
#
# Vector transformation:
#   Original: [1.000, 0.000, 0.000]
#   Rotated:  [0.000, 1.000, 0.000]
#
#   Original: [0.000, 1.000, 0.000]
#   Rotated:  [-0.000, 0.000, 1.000]

File: examples/attitude/rotation_matrix_operations.rs
//! Demonstrates rotation matrix operations including matrix multiplication
//! and vector transformations.

#[allow(unused_imports)]
use brahe as bh;
use nalgebra as na;

fn main() {
    // Create two rotation matrices
    // 90° rotation about X-axis
    let rm_x = bh::RotationMatrix::new(
        1.0, 0.0, 0.0,
        0.0, 0.0, -1.0,
        0.0, 1.0, 0.0
    ).unwrap();

    // 90° rotation about Z-axis
    let rm_z = bh::RotationMatrix::new(
        0.0, -1.0, 0.0,
        1.0, 0.0, 0.0,
        0.0, 0.0, 1.0
    ).unwrap();

    println!("Rotation matrix X (90° about X):");
    println!("  [{:.3}, {:.3}, {:.3}]", rm_x[(0, 0)], rm_x[(0, 1)], rm_x[(0, 2)]);
    println!("  [{:.3}, {:.3}, {:.3}]", rm_x[(1, 0)], rm_x[(1, 1)], rm_x[(1, 2)]);
    println!("  [{:.3}, {:.3}, {:.3}]", rm_x[(2, 0)], rm_x[(2, 1)], rm_x[(2, 2)]);

    println!("\nRotation matrix Z (90° about Z):");
    println!("  [{:.3}, {:.3}, {:.3}]", rm_z[(0, 0)], rm_z[(0, 1)], rm_z[(0, 2)]);
    println!("  [{:.3}, {:.3}, {:.3}]", rm_z[(1, 0)], rm_z[(1, 1)], rm_z[(1, 2)]);
    println!("  [{:.3}, {:.3}, {:.3}]", rm_z[(2, 0)], rm_z[(2, 1)], rm_z[(2, 2)]);

    // Matrix multiplication (compose rotations)
    // Apply rm_x first, then rm_z
    let rm_composed = rm_z * rm_x;
    println!("\nComposed rotation (X then Z):");
    println!("  [{:.3}, {:.3}, {:.3}]", rm_composed[(0, 0)], rm_composed[(0, 1)], rm_composed[(0, 2)]);
    println!("  [{:.3}, {:.3}, {:.3}]", rm_composed[(1, 0)], rm_composed[(1, 1)], rm_composed[(1, 2)]);
    println!("  [{:.3}, {:.3}, {:.3}]", rm_composed[(2, 0)], rm_composed[(2, 1)], rm_composed[(2, 2)]);

    // Transform a vector using rotation matrix
    // Rotate vector [1, 0, 0] by 90° about Z-axis using matrix multiplication
    let vector = na::SVector::<f64, 3>::new(1.0, 0.0, 0.0);
    let rotated = rm_z.to_matrix() * vector;  // Matrix-vector multiplication
    println!("\nVector transformation:");
    println!("  Original: [{:.3}, {:.3}, {:.3}]", vector[0], vector[1], vector[2]);
    println!("  Rotated:  [{:.3}, {:.3}, {:.3}]", rotated[0], rotated[1], rotated[2]);

    // Transform another vector
    let vector2 = na::SVector::<f64, 3>::new(0.0, 1.0, 0.0);
    let rotated2 = rm_z.to_matrix() * vector2;
    println!("\n  Original: [{:.3}, {:.3}, {:.3}]", vector2[0], vector2[1], vector2[2]);
    println!("  Rotated:  [{:.3}, {:.3}, {:.3}]", rotated2[0], rotated2[1], rotated2[2]);

    let eq_result = rm_x == rm_z;
    let neq_result = rm_x != rm_z;
    println!("\nEquality comparison:");
    println!("  rm_x == rm_z: {}", eq_result);
    println!("  rm_x != rm_z: {}", neq_result);
}

// Expected output:
// Rotation matrix X (90° about X):
//   [1.000, 0.000, 0.000]
//   [0.000, 0.000, -1.000]
//   [0.000, 1.000, 0.000]

// Rotation matrix Z (90° about Z):
//   [0.000, -1.000, 0.000]
//   [1.000, 0.000, 0.000]
//   [0.000, 0.000, 1.000]

// Composed rotation (X then Z):
//   [0.000, 0.000, 1.000]
//   [1.000, 0.000, 0.000]
//   [0.000, 1.000, 0.000]

// Vector transformation:
//   Original: [1.000, 0.000, 0.000]
//   Rotated:  [0.000, 1.000, 0.000]

//   Original: [0.000, 1.000, 0.000]
//   Rotated:  [-1.000, 0.000, 0.000]

// Equality comparison:
//   rm_x == rm_z: false
//   rm_x != rm_z: true
